src/Compilers/CSharp/Portable/Parser/LexerCache.cs

```diff
@@ -30,23 +29,144 @@ internal partial class LexerCache
+        private TextKeyedCache<SyntaxToken>? _tokenMap;
+        private CachingIdentityFactory<string, SyntaxKind>? _keywordKindMap;
         internal const int MaxKeywordLength = 10;
 
-        internal LexerCache()
+        private PooledStringBuilder? _stringBuilder;
+        private readonly char[] _identBuffer;
+        private SyntaxListBuilder? _leadingTriviaCache;
+        private SyntaxListBuilder? _trailingTriviaCache;
+
+        private const int LeadingTriviaCacheInitialCapacity = 128;
+        private const int TrailingTriviaCacheInitialCapacity = 16;
+
+        private LexerCache()
+        {
+            _identBuffer = new char[32];
+        }
+
+        public static LexerCache GetInstance()
         {
-            _triviaMap = TextKeyedCache<SyntaxTrivia>.GetInstance();
-            _tokenMap = TextKeyedCache<SyntaxToken>.GetInstance();
-            _keywordKindMap = s_keywordKindPool.Allocate();
+            return s_lexerCachePool.Allocate();
         }
 
-        internal void Free()
+        public void Free()
         {
-            _keywordKindMap.Free();
-            _triviaMap.Free();
-            _tokenMap.Free();
+            if (_keywordKindMap != null)
+            {
+                _keywordKindMap.Free();
+                _keywordKindMap = null;
+            }
+
+            if (_triviaMap != null)
+            {
+                _triviaMap.Free();
+                _triviaMap = null;
+            }
+
+            if (_tokenMap != null)
+            {
+                _tokenMap.Free();
+                _tokenMap = null;
+            }
+
+            if (_stringBuilder != null)
+            {
+                _stringBuilder.Free();
+                _stringBuilder = null;
+            }
+
+            if (_leadingTriviaCache != null)
+            {
+                if (_leadingTriviaCache.Capacity > LeadingTriviaCacheInitialCapacity * 2)
+                {
+                    // Don't reuse _leadingTriviaCache if it has grown too large for pooling.
+                    _leadingTriviaCache = null;
+                }
+                else
+                {
+                    _leadingTriviaCache.Clear();
+                }
+            }
+
+            if (_trailingTriviaCache != null)
+            {
+                if (_trailingTriviaCache.Capacity > TrailingTriviaCacheInitialCapacity * 2)
+                {
+                    // Don't reuse _trailingTriviaCache if it has grown too large for pooling.
+                    _trailingTriviaCache = null;
+                }
+                else
+                {
+                    _trailingTriviaCache.Clear();
+                }
+            }
+
+            s_lexerCachePool.Free(this);
+        }
+
+        internal char[] IdentBuffer => _identBuffer;
+
+        private TextKeyedCache<SyntaxTrivia> TriviaMap
+        {
+            get
+            {
+                _triviaMap ??= new TextKeyedCache<SyntaxTrivia>();
```

Comment:
>new TextKeyedCache<SyntaxTrivia>(); [](http://example.com/codeflow?start=31&length=35)

It looks like we are no longer allocating this instance from a pool. What is the rational behind this change?